app/services/authentication/authenticator.rb

```diff
@@ -0,0 +1,127 @@
+module Authentication
+  # Authenticator will perform one of these tree operations:
+  # 1. create a new user and match it to its authentication identity
+  # 1. update an existing user and align it to its authentication identity
+  # 1. return the current user if a user is given (already logged in scenario)
+  class Authenticator
```

Comment:
this is the old `AuthorizationService` but, hopefully, improved :D

Reply:
IMHO this is a candidate for (eventually) being replaced by a strategy pattern (and I don't say this lightly in a Ruby/Rails context). Make the decision early which one of the 3 cases we're dealing with and then call either `NewUserStrategy`, `UpdateUserStrategy` or `LoggedInUserStrategy`. I think the resulting three classes would be much easier to understand and they can still share methods by inhering from a basic `AuthStrategy`.

Reply:
Makes lots of sense! Love the strategy pattern! Will add a `TODO` note with your suggestion